thunder/dynamo/compiler.py

```diff
@@ -284,3 +289,107 @@ def last_backward_traces(self) -> list[Trace]:
+            # because input information is discarded once compilation is complete.
+            # We need this input information to infer the input range for each subgraph in the splitter.
+            profile_stats = ProfileStats(gm=copy.deepcopy(gm))
+            tao.id_to_profile_stats[idx] = profile_stats
+            tao.id_to_gm_map[idx] = gm
+
+            def record_stats(*args):
+                input_meta = input_to_example_input_meta(args)
+                tao.id_to_profile_stats[idx].input_meta_to_called_times[input_meta] += 1
+                return gm(*args)
+
+            tao.dispatch_map[idx] = record_stats
+
+            def _dispatch(*args):
+                return tao.dispatch_map[idx](*args)
+
+            return _dispatch
+
+        # NOTE When not in profiling mode, this just returns the gm for eager execution
+        return gm
+
+    cfn = torch.compile(fn, backend=dispatching_backend)
+
+    # Wraps the torch compiled callable so we can invalidate the profiling
+    #   callable for UX clarity
+    def profiling_callable(*args, **kwargs):
+        if tao.is_profiling:
+            return cfn(*args, **kwargs)
+
+        raise AssertionError(f"No longer profiling")
+
+    profiling_callable._compiled_fn = cfn
+    profiling_callable._tao = tao
+    return profiling_callable
+
+
+def thunder_optimize(
+    profiling_callable,
+    *,
+    gm_filter=default_filter,
+    optimizer=default_optimizer,
+):
+    """
+    Optimizes the given profiling callable based on the statistics collected during profiling.
+
+    Args:
+        profiling_callable: The callable returned by :func:`thunder.dynamo.compiler.thunder_profile`
+        gm_filter: Function that selects which graph modules to optimize
+        optimizer: Function that performs the actual optimization on each graph module
+
+    Returns:
+        The torch compiled function
+
+    Example:
+        >>> import torch
+        >>> from thunder.dynamo import thunder_profile
+        >>> def foo(x):
+        ...     return x + 1
+        >>> x = torch.randn(2, 2)
+        >>> pfoo = thunder_profile(foo)
+        >>> pfoo(x)  # Profile with first input
+        >>> stats = pfoo._tao.id_to_profile_stats  # Access collected statistics
+        >>> optfoo = thunder_optimize(pfoo)
+        >>> optfoo(x)  # Optimized with first input
+    """
+
+    # 1) Filters the FX graphs based on their statistics
+    indices_to_optimize: set[int] = gm_filter(profiling_callable)
+
+    # 2) Replaces the FX graphs that are to be optimized with the optimized
+    #      callables, and other graphs with the graph modules
+    tao = profiling_callable._tao
+    dispatch_map = tao.dispatch_map
+    id_to_gm_map = tao.id_to_gm_map
+    for idx, call in dispatch_map.items():
+        if idx in indices_to_optimize:
+            gm = id_to_gm_map[idx]
+            profile_stats = tao.id_to_profile_stats[idx]
+            dispatch_map[idx] = optimizer(gm, profile_stats)
```

Comment:
If we let `optimizer` throw a `NotImplementedError` (see below), then a slight expansion of this might be:

```
try: 
  dispatch_map[idx] = optimizer(gm, profile_stats)
catch NotImplementedError:
  # Executes the gm eagerly if the optimizer can't optimize it
  dispatch_map[idx] = gm
```